core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.execution.scheduler.PipelinedStageExecution.State.FLUSHING;
+import static io.trino.execution.scheduler.PipelinedStageExecution.State.RUNNING;
+import static io.trino.execution.scheduler.PipelinedStageExecution.State.SCHEDULED;
+import static io.trino.sql.planner.plan.AggregationNode.Step.FINAL;
+import static io.trino.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static io.trino.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+/**
+ * Schedules stages in sequence in order to provide the best resource utilization.
+ * This means that stages which output won't be consumed (e.g. join probe side) will
+ * not be scheduled until dependent stages finish (e.g. join build source stages).
+ * Contrary to {@link PhasedExecutionPolicy}, {@link SequentialExecutionSchedule} will
+ * schedule multiple source stages in order to fully utilize IO.
+ */
+public class SequentialExecutionSchedule
+        implements ExecutionSchedule
+{
+    /**
+     * Graph representing a before -> after relationship between fragments.
+     * Destination fragment should be started only when source stage is completed.
+     */
+    private final DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentDependency;
+    /**
+     * Graph representing topology between fragments (e.g. child -> parent relationship).
+     */
+    private final DirectedGraph<PlanFragmentId, FragmentsEdge> fragmentTopology;
+    private final Map<PlanFragmentId, StageExecution> stagesByFragmentId;
+    private final Set<StageExecution> activeSources = new HashSet<>();
+
+    @GuardedBy("this")
+    private SettableFuture<Void> rescheduleFuture = SettableFuture.create();
+
+    public static SequentialExecutionSchedule forStages(Collection<PipelinedStageExecution> stages)
+    {
+        return new SequentialExecutionSchedule(stages.stream()
+                .map(SequentialExecutionSchedule::createStageExecution)
+                .collect(toImmutableSet()));
+    }
+
+    @VisibleForTesting
+    SequentialExecutionSchedule(Collection<StageExecution> stages)
+    {
+        fragmentDependency = new DefaultDirectedGraph<>(new FragmentsEdgeFactory());
+        fragmentTopology = new DefaultDirectedGraph<>(new FragmentsEdgeFactory());
+        stagesByFragmentId = stages.stream()
+                .collect(toImmutableMap(stage -> stage.getFragment().getId(), identity()));
+        extractDependencies(
```

Comment:
It is not obvious what does this method return. How about creating a hash set and passing it as a parameter here?